File: sdks/node/src/dashboard/urlSafety.ts
// Open-redirect guard for post-login redirect targets.

/**
 * Whether `path` is safe as a same-origin redirect target. Accepts only
 * relative paths rooted at `/`; rejects absolute URLs, protocol-relative
 * URLs (`//evil.com/x`), and backslash tricks. Empty input is rejected so
 * callers fall back to a default explicitly.
 */
export function isSafeRedirect(path: string | null | undefined): boolean {
  if (!path?.startsWith("/")) {
    return false;
  }
  if (path.startsWith("//") || path.startsWith("/\\")) {
    return false;
  }
  // A parseable absolute URL (with scheme or authority) is never safe.
  try {
    // Base-relative parse: if the path smuggles a scheme/host, the resolved
    // origin differs from the sentinel base.
    const resolved = new URL(path, "http://taskito.invalid");
    return resolved.origin === "http://taskito.invalid";
  } catch {
    return false;
  }
}

File: sdks/java/src/main/java/org/byteveda/taskito/dashboard/auth/oauth/UrlSafety.java
package org.byteveda.taskito.dashboard.auth.oauth;

import java.net.URI;

/**
 * Open-redirect guard for the post-login {@code next} target.
 *
 * <p>Only same-origin relative paths rooted at {@code /} are accepted. Absolute
 * URLs ({@code https://evil.com}), protocol-relative URLs ({@code //evil.com}),
 * and backslash tricks ({@code /\evil}) are rejected so an attacker can't craft
 * a login link that bounces the browser off-site after authentication.
 */
public final class UrlSafety {

    private UrlSafety() {}

    /** Whether {@code path} is safe to use as a same-origin post-login redirect. */
    public static boolean isSafeRedirect(String path) {
        if (path == null || path.isEmpty()) {
            return false;
        }
        if (!path.startsWith("/")) {
            return false;
        }
        // "//host" is protocol-relative; "/\host" is a browser-normalised variant.
        if (path.startsWith("//") || path.startsWith("/\\")) {
            return false;
        }
        URI uri;
        try {
            uri = URI.create(path);
        } catch (IllegalArgumentException e) {
            return false;
        }
        // A bare rooted path has neither a scheme nor an authority/host.
        return uri.getScheme() == null && uri.getAuthority() == null && uri.getHost() == null;
    }
}
